packages/react-jss/src/JssProvider.js

```diff
@@ -20,85 +13,80 @@ type Props = {
-    } = this.props
-    // eslint-disable-next-line react/react/destructuring-assignment
-    const sheetOptions = this.context[ns.sheetOptions] || {}
-    const context: Context = {[ns.sheetOptions]: sheetOptions}
+  managers: Managers = {}
+
+  /**
+   * We need to merge the outer context with the props,
+   * because we allow overriding any prop at any level.
+   */
+  createContext(outerContext: Context): Context {
+    const {registry, classNamePrefix, jss, generateId, disableStylesGeneration, media} = this.props
+    // Clone the outer context
+    const context = {
+      ...outerContext,
+      managers: this.managers
+    }
 
     if (registry) {
-      context[ns.sheetsRegistry] = registry
-      // This way we identify a new request on the server, because user will create
-      // a new Registry instance for each.
-      if (registry !== this.registry) {
-        // We reset managers because we have to regenerate all sheets for the new request.
-        this.managers = {}
-        this.registry = registry
-      }
+      context.registry = registry
     }
 
-    // Make sure we don't loose managers when JssProvider is used inside of a stateful
-    // component which decides to rerender.
-    context[ns.managers] = this.managers
-
     if (generateId) {
-      sheetOptions.generateId = generateId
-    } else if (!sheetOptions.generateId) {
+      context.sheetOptions.generateId = generateId
+    } else if (!context.sheetOptions.generateId) {
       if (!this.generateId) {
-        let createGenerateId = createGenerateIdDefault
-        if (localJss && localJss.options.createGenerateId) {
-          createGenerateId = localJss.options.createGenerateId
-        }
-        // Make sure we don't loose the generator when JssProvider is used inside of a stateful
-        // component which decides to rerender.
         this.generateId = createGenerateId()
       }
+      context.sheetOptions.generateId = this.generateId
+    }
+
+    // Merge the classname prefix
+    if (classNamePrefix) {
+      context.sheetOptions.classNamePrefix =
+        (context.sheetOptions.classNamePrefix || '') + classNamePrefix
+    }
 
-      sheetOptions.generateId = this.generateId
+    if (media !== undefined) {
+      context.sheetOptions.media = media
+    }
+
+    if (jss) {
+      context.jss = jss
     }
 
-    if (classNamePrefix) sheetOptions.classNamePrefix = classNamePrefix
-    if (localJss) context[ns.jss] = localJss
     if (disableStylesGeneration !== undefined) {
-      sheetOptions.disableStylesGeneration = disableStylesGeneration
+      context.disableStylesGeneration = disableStylesGeneration
     }
 
     return context
   }
 
-  registry: SheetsRegistry
+  generateId: GenerateId
 
-  managers: {}
+  renderProvider = (outerContext: Context) => {
+    const {children} = this.props
 
-  generateId: GenerateId
+    return (
+      <JssContext.Provider value={this.createContext(outerContext)}>{children}</JssContext.Provider>
```

Comment:
what should user should know from changelog in this regard?

Reply:
That's just an internal change. Also mentioned the switch to the new React context in the changelog